File: python/leetcode/282.py
class Solution:
    def addOperators(self, num, target):
        res = []
        for i in range(1, len(num) + 1):
            # prevent "00*" as a number
            if i == 1 or (i > 1 and num[0] != "0"):
                # this step put first number in the string
                self.dfs(target, num[i:], num[:i], int(num[:i]), int(num[:i]), res)
        return res

    def dfs(self, target, num, ans, cur, last, res):
        if not num:
            if cur == target:
                res.append(ans)
            return
        for i in range(1, len(num) + 1):
            val = num[:i]
            # prevent "00*" as a number
            if i == 1 or (i > 1 and num[0] != "0"):
                self.dfs(target, num[i:], ans + "+" + val,
                         cur + int(val), int(val), res)
                self.dfs(target, num[i:], ans + "-" + val,
                         cur - int(val), -int(val), res)
                self.dfs(target, num[i:], ans + "*" + val, cur - last +
                         last * int(val), last * int(val), res)

File: cpp/leetcode/282.cpp
class Solution {
  private:
    void dfs(int target, string &num, string ans, int cur, int last, vector<int>& res) {
        if (nums.empty()) {
            if (cur == target) {
                res.push_back(ans);
            }
            return;
        }
        for (int i = 1; i <= num.size(); i++) {
            string t = num.substr(0, i);
            long now = stol(t);
            if (to_string(now).size() != t.size()) {
                continue;
            }
            dfs(target, num.substr(i), ans + '+' + t, cur + now, now, res);
            dfs(target, num.substr(i), ans + '-' + t, cur - now, -now, res);
            dfs(target, num.substr(i), ans + '*' + t, cur - last + last * now, last * now, res);
        }
    }

  public:
    vector<string> addOperators(string num, int target) {
        vector<string> res;
        if (num.empty()) {
            return res;
        }
        for (int i = 1; i <= num.size(); i++) {
            string s = num.substr(0, i);
            long ans = stol(s);
            if (to_string(ans).size() != s.size()) {
                continue;
            }
            dfs(res, num, target, s, i, ans, ans, '#');
        }
        return res;
    }
};